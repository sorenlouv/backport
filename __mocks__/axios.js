const axios = jest.genMockFromModule('axios');
axios.get = jest.fn(() => {
  throw new Error('Forgot to mock axios.get');
});
axios.post = jest.fn(() => {
  throw new Error('Forgot to mock axios.post');
});
module.exports = axios;
