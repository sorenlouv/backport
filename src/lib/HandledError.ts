export class HandledError extends Error {
  constructor(message: string) {
    super(message);
    Error.captureStackTrace(this, HandledError);
    this.name = 'HandledError';
  }
}

export function printHandledError(e: HandledError, { rethrow = true } = {}) {
  if (e.name === 'HandledError') {
    console.error(e.message);
  } else {
    console.error(e);
    if (rethrow) {
      throw e;
    }
  }
}
