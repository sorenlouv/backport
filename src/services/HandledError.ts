import { Commit } from './sourceCommit/parseSourceCommit';

type ErrorContext =
  | {
      code: 'merge-conflict-exception';
      conflictingFiles: string[];
      commitsWithoutBackports: {
        formatted: string;
        commit: Commit;
      }[];
    }
  | {
      code: 'custom-exception';
      message: string;
    }
  | {
      code: 'no-branches-exception' | 'abort-exception';
    };

function getMessage(errorContext: ErrorContext): string {
  switch (errorContext.code) {
    case 'merge-conflict-exception':
      return `Commit could not be cherrypicked due to conflicts in: ${errorContext.conflictingFiles.join(
        ','
      )}`;
    case 'no-branches-exception':
      return 'There are no branches to backport to. Aborting.';
    case 'abort-exception':
      return 'Aborted';
    case 'custom-exception':
      return errorContext.message;
  }
}

export class HandledError extends Error {
  errorContext: ErrorContext;
  constructor(errorContextOrString: ErrorContext | string) {
    const errorContext: ErrorContext =
      typeof errorContextOrString === 'string'
        ? { code: 'custom-exception', message: errorContextOrString }
        : errorContextOrString;
    const message = getMessage(errorContext);
    super(message);
    Error.captureStackTrace(this, HandledError);
    this.name = 'HandledError';
    this.message = message;
    this.errorContext = errorContext;
  }
}
