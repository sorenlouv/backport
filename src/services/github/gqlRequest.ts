import axios, { AxiosError } from 'axios';
import { HandledError } from '../HandledError';
import { logger } from '../logger';

interface GithubResponse<DataResponse> {
  data: DataResponse;
  errors: {
    type: string;
    path: string[];
    locations: {
      line: number;
      column: number;
    }[];
    message: string;
  }[];
}

export async function gqlRequest<DataResponse>({
  githubApiBaseUrlV4,
  query,
  variables,
  accessToken,
}: {
  githubApiBaseUrlV4: string;
  query: string;
  variables?: {
    [key: string]: string | number | null;
  };
  accessToken: string;
}) {
  try {
    logger.verbose('gql query:', query);
    logger.verbose('gql variables:', variables);
    const response = await axios.post<GithubResponse<DataResponse>>(
      githubApiBaseUrlV4,
      { query, variables },
      {
        headers: {
          'Content-Type': 'application/json',
          Authorization: `bearer ${accessToken}`,
        },
      }
    );

    logger.debug('gql response:', response.data);

    if (response.data.errors) {
      const newError = new Error();
      //@ts-ignore
      newError.response = response;
      throw newError;
    }

    return response.data.data;
  } catch (ex) {
    const e = ex as AxiosError<{ errors: { message: string }[] | undefined }>;
    logger.info(e.message);

    if (e.response?.data) {
      logger.info('API v4 config:', e.config);
      logger.info('API v4 response headers:', e.response.headers);
      logger.info('API v4 response data', e.response.data);

      const errorMessages = e.response.data.errors?.map(
        (error) => error.message
      );

      if (errorMessages) {
        throw new HandledError(errorMessages.join(', '));
      }

      throw new HandledError(
        `Unexpected response from Github:\n${JSON.stringify(
          e.response.data,
          null,
          2
        )}`
      );
    }

    throw e;
  }
}
