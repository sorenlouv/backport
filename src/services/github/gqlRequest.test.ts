import axios from 'axios';
import { HandledError } from '../HandledError';
import { gqlRequest } from './gqlRequest';
import dedent from 'dedent';

describe('gqlRequest', () => {
  describe('when request succeeds', () => {
    let spy: jest.SpyInstance;
    beforeEach(() => {
      spy = jest.spyOn(axios, 'post').mockResolvedValue({
        data: {
          data: 'some data'
        }
      } as any);
    });

    it('should return correct response', async () => {
      expect(
        await gqlRequest({
          accessToken: 'myAccessToken',
          apiHostname: 'myApiHostname',
          query: 'myQuery',
          variables: {
            foo: 'bar'
          }
        })
      ).toEqual('some data');
    });

    it('should call with correct args', async () => {
      expect(spy).toHaveBeenCalledWith(
        'https://myApiHostname/graphql',
        {
          query: 'myQuery',
          variables: { foo: 'bar' }
        },
        {
          headers: {
            Authorization: 'bearer myAccessToken',
            'Content-Type': 'application/json'
          }
        }
      );
    });
  });

  describe('when request fails', () => {
    beforeEach(() => {
      jest.spyOn(axios, 'post').mockRejectedValue({
        response: {
          data: {
            errors: [{ message: 'some error' }, { message: 'some other error' }]
          }
        }
      } as any);
    });

    it('should return parsed github error', async () => {
      return expect(
        gqlRequest({
          accessToken: 'myAccessToken',
          apiHostname: 'myApiHostname',
          query: 'myQuery',
          variables: {
            foo: 'bar'
          }
        })
      ).rejects.toThrowError(
        new HandledError(
          dedent(`Unexpected response from Github:

          some error, some other error`)
        )
      );
    });
  });
});
