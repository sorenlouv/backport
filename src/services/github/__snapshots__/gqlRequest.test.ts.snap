// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gqlRequest when request fails without error messages should return parsed github error 1`] = `
"Unexpected response from Github:
{
  \\"foo\\": \\"bar\\"
}"
`;
